### PR TITLE
feat: TestCase extractor tool

### PR DESCRIPTION
**Pull request by jiri-sotona**
A tool for extraction of FEEL related  test cases from TCK test suite

Referring to https://github.com/dmn-tck/tck

### CODE REVIEW

File: internal/testcase-extractor/main.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
+package main
```
Comment (richardbizik) [inline internal/testcase-extractor/main.go:1]: please put this file into ./cmd/testcase-extractor/main.go 
In go the main classes for binaries are usually found in cmd/<binary>/main.go

**Review by richardbizik — CHANGES_REQUESTED** [review]

File: internal/testcase-extractor/main.go (hunk 1 above)
Comment (jiri-sotona) [reply internal/testcase-extractor/main.go]: Done.

File: cmd/testcase-extractor/main.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+func printError(format string, a ...any) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+func main() {
+	startTs := time.Now()
+
+	dir := flag.String("dir", ".", "Directory to start search for DMN model files and Test Cases files.")
```
Comment (nitram509) [inline cmd/testcase-extractor/main.go:252]: It took me some time to figure out what I need to do, in order to run it locally.
Please add a README file, to explain basic usage and the preconditions (e.g. downloading the TCK, etc.)

**Review by nitram509 — CHANGES_REQUESTED** [review]

File: cmd/testcase-extractor/main.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	XMLName     xml.Name             `xml:"testCase"`
+	Id          *string              `xml:"id,attr"`
+	Description *string              `xml:"description"`
+	Results     []*TckTestCaseResult `xml:"resultNode"`
+}
+
+type TckTestCaseResult struct {
+	XMLName  xml.Name      `xml:"resultNode"`
+	Name     *string       `xml:"name,attr"`
+	Expected NilableString `xml:"expected>value"`
```
Comment (richardbizik) [inline cmd/testcase-extractor/main.go:42]: Why is this not a pointer to string but a custom type?

**Review by richardbizik — CHANGES_REQUESTED** [review]
I would be expecting a test that runs these test cases as a part of this PR.
Could be disabled by default but it should be here as a proof that the extractor is working.

File: cmd/testcase-extractor/main.go [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+func main() {
+	startTs := time.Now()
+
+	dir, outputFilename, forceStats := parseFlags()
+
+	searchDir, err := filepath.Abs(*dir)
+	if err != nil {
+		printError("Error getting search dir absolute path: %v\n", err)
+		panic(err)
```
Comment (richardbizik) [inline cmd/testcase-extractor/main.go:354]: usually you don`t have to panic right after printing the error and you can just os.Exit(1) in the main function

File: cmd/testcase-extractor/main.go [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+		outputFilenameAbs, _ := filepath.Abs(outputFile.Name())
+		outputFileInfo = fmt.Sprintf(
+			" to file "+
+				"\n\t%v",
+			outputFilenameAbs,
+		)
+	}
+
+	fmt.Printf(
+		"\n"+
```
Comment (richardbizik) [inline cmd/testcase-extractor/main.go:333]: fyi for more complicated templating go has [text/template](https://pkg.go.dev/text/template) (you dont have to change this now)

**Review by richardbizik — CHANGES_REQUESTED** [review]

File: cmd/testcase-extractor/main.go (hunk 3 above)
Comment (jiri-sotona) [reply cmd/testcase-extractor/main.go:42]: With `*string` type the `xml.Unmarshal(...)` gives empty string ("") for `xs:nil` values. We need explicit `null` value in YAML because test case mandates that `null` is the result of expression evaluation.

File: cmd/testcase-extractor/README.adoc [hunk 6]
```diff
@@ -0,0 +1,67 @@
+= `testcase-extractor`
```
[the remark this answers is not in the document]
Comment (jiri-sotona) [reply cmd/testcase-extractor/README.adoc:1]: I see. I have no problem with that. Personally, I prefer Asciidoc because is more feature rich. But it makes sense to keep all docs in same format.

File: cmd/testcase-extractor/main.go (hunk 3 above)
Comment (richardbizik) [reply cmd/testcase-extractor/main.go:42]: Oh I see, thanks for the explanation.

File: cmd/testcase-extractor/main.go (hunk 4 above)
Comment (jiri-sotona) [reply cmd/testcase-extractor/main.go]: Changed as suggested.

File: cmd/testcase-extractor/main.go (hunk 2 above)
Comment (jiri-sotona) [reply cmd/testcase-extractor/main.go]: Done.

File: cmd/testcase-extractor/Makefile [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# Variables
+BINARY_NAME=testcase-extractor
```
Comment (richardbizik) [inline cmd/testcase-extractor/Makefile:2]: we don't have to build the binary it will be enough to run it by `go run cmd/...`

File: cmd/testcase-extractor/Makefile [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+.PHONY: clean-tck
+clean-tck:
+	@rm -rf $(TCK_DIR)
+
+.PHONY: extract-testcases-only
+extract-testcases-only: build
+	$(BUILD_DIR)/$(BINARY_NAME) --dir $(TCK_DIR) --output-file $(TEST_CONFIG_FILE)
+
+.PHONY: extract-testcases
```
Comment (richardbizik) [inline cmd/testcase-extractor/Makefile:56]: I would merge all the targets to clone, run, clean into this one once the makefile is moved

**Review by richardbizik — CHANGES_REQUESTED** [review]
Please move the makefile into the root of the project.

File: cmd/testcase-extractor/Makefile (hunk 8 above)
Comment (jiri-sotona) [reply cmd/testcase-extractor/Makefile]: Done.

File: cmd/testcase-extractor/Makefile (hunk 7 above)
Comment (jiri-sotona) [reply cmd/testcase-extractor/Makefile]: Done.

File: cmd/testcase-extractor/README.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (jiri-sotona) [reply cmd/testcase-extractor/README.adoc]: Done.

**Comment by jiri-sotona** [comment]
> > Please move the makefile into the root of the project.
> 
> My impression was that we want to use this tool temporarily to extract and run FEEL tests until complete TCK test suite will be run the way we discussed on last meeting. Or?

Tool will stay. Moved taget to Makefile in repo root.

**Review by richardbizik — APPROVED** [review]
